src/test/java/org/apache/maven/plugins/clean/CleanerTest.java

```diff
@@ -0,0 +1,173 @@
+        final Set<PosixFilePermission> initialPermissions = getPosixFilePermissions(basedir);
+        final String rwxrwxr_x = PosixFilePermissions.toString(initialPermissions);
+        // Prevent directory listing, which will result in a DirectoryNotEmptyException.
+        final String rw_rw_r__ = rwxrwxr_x.replace('x', '-');
+        final Set<PosixFilePermission> permissions = PosixFilePermissions.fromString(rw_rw_r__);
+        setPosixFilePermissions(basedir, permissions);
+        try {
+            final Cleaner cleaner = new Cleaner(null, log, false, null, null);
+            final IOException exception =
+                    assertThrows(IOException.class, () -> cleaner.delete(basedir.toFile(), null, false, true, true));
+            assertEquals("Failed to delete " + basedir, exception.getMessage());
+            final DirectoryNotEmptyException cause =
+                    assertInstanceOf(DirectoryNotEmptyException.class, exception.getCause());
+            assertEquals(basedir.toString(), cause.getMessage());
+        } finally {
+            // Allow the tempDir to be cleared by the @TempDir extension.
+            setPosixFilePermissions(basedir, initialPermissions);
+        }
+    }
+
+    @Test
+    void deleteLogsWarningWithoutRetryWhenNoPermission(@TempDir Path tempDir) throws Exception {
+        assumeTrue(POSIX_COMPLIANT);
+        when(log.isWarnEnabled()).thenReturn(true);
+        final Path basedir = createDirectory(tempDir.resolve("target")).toRealPath();
+        final Path file = createFile(basedir.resolve("file"));
+        final Set<PosixFilePermission> initialPermissions = getPosixFilePermissions(basedir);
+        final String rwxrwxr_x = PosixFilePermissions.toString(initialPermissions);
+        final String r_xr_xr_x = rwxrwxr_x.replace('w', '-');
+        final Set<PosixFilePermission> permissions = PosixFilePermissions.fromString(r_xr_xr_x);
+        setPosixFilePermissions(basedir, permissions);
+        try {
+            final Cleaner cleaner = new Cleaner(null, log, false, null, null);
+            assertDoesNotThrow(() -> cleaner.delete(basedir.toFile(), null, false, false, false));
+            verify(log, times(2)).warn(any(CharSequence.class), any(Throwable.class));
+            InOrder inOrder = inOrder(log);
+            ArgumentCaptor<AccessDeniedException> cause1 = ArgumentCaptor.forClass(AccessDeniedException.class);
+            inOrder.verify(log).warn(eq("Failed to delete " + file), cause1.capture());
+            assertEquals(file.toString(), cause1.getValue().getMessage());
+            ArgumentCaptor<DirectoryNotEmptyException> cause2 =
+                    ArgumentCaptor.forClass(DirectoryNotEmptyException.class);
+            inOrder.verify(log).warn(eq("Failed to delete " + basedir), cause2.capture());
+            assertEquals(basedir.toString(), cause2.getValue().getMessage());
+        } finally {
+            setPosixFilePermissions(basedir, initialPermissions);
```

Comment:
why do you need to reset this here? Feels like the object shouldn't be used again

Reply:
To avoid resource exhaustion. It's good practice to reset persistent state which is modified in a test. The files I make in this test are persistent, and I rely on the `@TempDir` extension in the JUnit framework to delete them. If I leave permissions as they are, i.e. deliberately restricted so that files cannot be deleted, I imagine it could be that deleting the directory fails, resulting not per se in an exception but in lingering temporary directories. Over time, lingering temporary directories can cause "no space left on device" on test systems. By resetting the permissions, I reset the persistent state, allowing the JUnit framework to clear the temporary directory, avoiding resource exhaustion.

Reply:
Something's off with that. Test resources including files should not be shared between test methods. Per Junit docs "The temporary directory will be shared by all tests in a class when the annotation is present on a static field or on a parameter of a [@BeforeAll](https://junit.org/junit5/docs/5.4.1/api/org/junit/jupiter/api/BeforeAll.html) method. Otherwise — for example, when @TempDir is only used on instance fields or on parameters in test, [@BeforeEach](https://junit.org/junit5/docs/5.4.1/api/org/junit/jupiter/api/BeforeEach.html), or [@AfterEach](https://junit.org/junit5/docs/5.4.1/api/org/junit/jupiter/api/AfterEach.html) methods — each test will use its own temporary directory."